Terminal.Gui/Text/TextFormatter.cs

```diff
@@ -747,8 +691,236 @@ public List<string> GetLines ()
+    /// <remarks>
+    ///     <para>
+    ///         Does not return the size of the formatted text but the size that will be used to constrain the text when
+    ///         formatted.
+    ///     </para>
+    ///     <para>
+    ///         Use <see cref="FormatAndGetSize"/> to get the size of the formatted text.
+    ///     </para>
+    ///     <para>When set, <see cref="NeedsFormat"/> is set to <see langword="true"/>.</para>
+    /// </remarks>
+    public Size? Size
+    {
+        get
+        {
+            if (_width is null || _height is null)
+            {
+                return null;
+            }
+
+            return new Size (_width.Value, _height.Value);
+        }
+        set
+        {
+            if (AutoSize)
+            {
+                Size size = EnableNeedsFormat (GetAutoSize ());
+                _width = size.Width;
+                _height = size.Height;
+            }
+            else
+            {
+                if (value is null)
+                {
+                    _width = null;
+                    _height = null;
+                    EnableNeedsFormat (true);
+                }
+                else
+                {
+                    Size size = EnableNeedsFormat (value.Value);
+                    _width = size.Width;
+                    _height = size.Height;
+                }
+            }
+        }
+    }
+
+    /// <summary>Gets or sets the number of columns used for a tab.</summary>
+    public int TabWidth
+    {
+        get => _tabWidth;
+        set => _tabWidth = EnableNeedsFormat (value);
+    }
+
+    /// <summary>The text to be formatted. This string is never modified.</summary>
+    public string Text
+    {
+        get => _text!;
+        set
+        {
+            _text = EnableNeedsFormat (value);
+
+            if (AutoSize)
+            {
+                Size = GetAutoSize ();
+                ;
+            }
+        }
+    }
+
+    /// <summary>Gets or sets the vertical text-alignment.</summary>
+    /// <value>The text vertical alignment.</value>
+    public Alignment VerticalAlignment
+    {
+        get => _textVerticalAlignment;
+        set => _textVerticalAlignment = EnableNeedsFormat (value);
+    }
+
+    /// <summary>Gets or sets the width <see cref="Text"/> will be constrained to when formatted.</summary>
+    /// <remarks>
+    ///     <para>
+    ///         Does not return the width of the formatted text but the width that will be used to constrain the text when
+    ///         formatted.
+    ///     </para>
+    ///     <para>
+    ///         Use <see cref="FormatAndGetSize"/> to get the size of the formatted text.
+    ///     </para>
+    ///     <para>When set, <see cref="NeedsFormat"/> is set to <see langword="true"/>.</para>
+    /// </remarks>
+    public int? Width
+    {
+        get => _width;
+        set
+        {
+            if (_width == value)
+            {
+                return;
+            }
+
+            if (value < 0)
```

Comment:
This is more expressive and helps static analysis:

```csharp
ArgumentOutOfRangeException.ThrowIfNegative (value.GetValueOrDefault (), nameof (Width));
```

It also points out a problem with nullable value types. If you try to just pass `value` to that throw helper, it's a compile error, because nullable value types suck. They do not implement any of the interfaces of T. This also means that Width and Height cannot be used by anything else that wants to use any of those interfaces without a null check and cast (such as via the `GetValueOrDefault()` method).